dynamiqs/sparse.py

```diff
@@ -0,0 +1,149 @@
+import functools
+
+import jax
+import jax.numpy as jnp
+from jaxtyping import Array
+
+
+class SparseDIA:
+    def __init__(self, diags: Array, offsets: tuple[int]):
+        self.offsets = offsets
+        self.diags = diags
+
+
+    def to_dense(self) -> Array:
+        """Turn any set of diagonals & offsets in a full NxN matrix."""
+        N = self.diags.shape[1]
+        out = jnp.zeros((N, N))
+        for offset, diag in zip(self.offsets, self.diags):
+            start = max(0, offset)
+            end = min(N, N + offset)
+            out += jnp.diag(diag[start:end], k=offset)
+        return out
+    
+    def to_sparse(self, matrix) -> Array:
+        """Turn a NxN sparse matrix into the amazing SparseDIA format"""
+        diagonals_with_offsets = {}
+        
+        # there are 2N-1 offsets in a NxN matrix
+        offset_range = 2 * matrix.shape[0] - 1
+        offset_center = offset_range // 2
+
+        for offset in range(-offset_center, offset_center + 1):
+            diagonal = jnp.diagonal(matrix, offset=offset)
+            if jnp.any(diagonal != 0):
+                diagonals_with_offsets[offset] = diagonal
+
+        return diagonals_with_offsets
```

Comment:
This function is a class creator, so it should be independent from the class. In other words, the `self` object is never used.